### ISSUE TITLE
GifFrameLoader: NPE when accessing android.graphics.Bitmap.getWidth()

### ISSUE THREAD

**Issue by realdeadfish** [bumptech/glide#2231]
In a `RecyclerView` implementation, we clear up a previously loaded GIF image when the view gets recycled.

**Glide Version**: 4.0.0
**Integration libraries**: OkHttp
**Device/Android Version**: OnePlus 3T

**Code:**

```java
 Glide.with(rootView.getContext()).clear(holder.imageView);
 holder.imageView.setVisibility(View.INVISIBLE);
```

**Stack trace / LogCat**:
```ruby
E  java.lang.NullPointerException: Attempt to invoke virtual method 'int android.graphics.Bitmap.getWidth()' on a null object reference
E      at com.bumptech.glide.load.resource.gif.GifFrameLoader.getWidth(GifFrameLoader.java:127)
E      at com.bumptech.glide.load.resource.gif.GifDrawable.getIntrinsicWidth(GifDrawable.java:215)
E      at android.widget.ImageView.updateDrawable(ImageView.java:963)
E      at android.widget.ImageView.setImageDrawable(ImageView.java:526)
E      at com.bumptech.glide.request.target.ImageViewTarget.setDrawable(ImageViewTarget.java:44)
E      at com.bumptech.glide.request.target.ImageViewTarget.onLoadCleared(ImageViewTarget.java:83)
E      at com.bumptech.glide.request.SingleRequest.clear(SingleRequest.java:272)
E      at com.bumptech.glide.manager.RequestTracker.clearRemoveAndRecycle(RequestTracker.java:62)
E      at com.bumptech.glide.RequestManager.untrack(RequestManager.java:437)
E      at com.bumptech.glide.Glide.removeFromManagers(Glide.java:634)
E      at com.bumptech.glide.RequestManager.untrackOrDelegate(RequestManager.java:426)
E      at com.bumptech.glide.RequestManager.clear(RequestManager.java:412)
E      at com.bumptech.glide.RequestManager.clear(RequestManager.java:397)
```

**Comment by realdeadfish** [comment]
I think it might be related how I handle the transition of a previous image on load. I.e. I do

```java
RequestOptions options = new RequestOptions();
if (holder.imageView.getVisibility() == View.VISIBLE) {
    options.placeholder(holder.imageView.getDrawable().mutate());
}
Glide.with(rootView)
     .load(file)
     .apply(options)
     .listener(animateAppearance(...))
     .into(holder.imageView);
```

to have a cross-fade between the currently loaded thumbnail and the new image. Now when the image is cleared, it apparently tries to set the original placeholder drawable and therefor runs into the above error.

Note that just try-catching this issue makes matters worse, because the next load on the same `ImageView` will crash then like this:

```ruby
E  java.lang.NullPointerException: Attempt to invoke virtual method 'int android.graphics.Bitmap.getWidth()' on a null object reference
E      at com.bumptech.glide.load.resource.gif.GifFrameLoader.getWidth(GifFrameLoader.java:127)
E      at com.bumptech.glide.load.resource.gif.GifDrawable.getIntrinsicWidth(GifDrawable.java:215)
E      at android.widget.ImageView.invalidateDrawable(ImageView.java:239)
E      at android.graphics.drawable.Drawable.invalidateSelf(Drawable.java:435)
E      at android.graphics.drawable.Drawable.setBounds(Drawable.java:214)
E      at android.widget.ImageView.configureBounds(ImageView.java:1181)
E      at android.widget.ImageView.setFrame(ImageView.java:1159)
E      at android.view.View.layout(View.java:17742)
E      at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
E      at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
E      at android.view.View.layout(View.java:17745)
E      at android.view.ViewGroup.layout(ViewGroup.java:5579)
E      at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
E      at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
E      at android.view.View.layout(View.java:17745)
E      at android.view.ViewGroup.layout(ViewGroup.java:5579)
E      at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1079)
E      at android.view.View.layout(View.java:17745)
E      at android.view.ViewGroup.layout(ViewGroup.java:5579)
E      at android.support.v7.widget.RecyclerView$LayoutManager.layoutDecoratedWithMargins(RecyclerView.java:8839)
E      at android.support.v7.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1609)
E      at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1511)
E      at android.support.v7.widget.LinearLayoutManager.scrollBy(LinearLayoutManager.java:1325)
E      at android.support.v7.widget.LinearLayoutManager.scrollVerticallyBy(LinearLayoutManager.java:1061)
E      at android.support.v7.widget.RecyclerView$ViewFlinger.run(RecyclerView.java:4726)
E      at android.view.Choreographer$CallbackRecord.run(Choreographer.java:874)
E      at android.view.Choreographer.doCallbacks(Choreographer.java:686)
E      at android.view.Choreographer.doFrame(Choreographer.java:618)
E      at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:860)
E      at android.os.Handler.handleCallback(Handler.java:751)
E      at android.os.Handler.dispatchMessage(Handler.java:95)
E      at android.os.Looper.loop(Looper.java:154)
```

I guess this is the dangling placeholder item that we were missing cleaning up there.

**Comment by realdeadfish** [comment]
On a related note, if I try-catch the clear and then call `holder.imageView.setImageDrawable(null)` myself afterwards, I still receive a crash on next load:

```java
E  Process: com.xing.android, PID: 24680
E  java.lang.IllegalStateException: Failed to remove target from managers
E      at com.bumptech.glide.Glide.removeFromManagers(Glide.java:639)
E      at com.bumptech.glide.RequestManager.untrackOrDelegate(RequestManager.java:426)
E      at com.bumptech.glide.RequestManager.clear(RequestManager.java:412)
E      at com.bumptech.glide.RequestBuilder.into(RequestBuilder.java:359)
E      at com.bumptech.glide.RequestBuilder.into(RequestBuilder.java:413)
```

**Comment by sjudd** [comment]
It's unsafe to set a Drawable loaded from Glide as a placeholder like that. Glide's resources for any previous loads are recycled as soon as you call into() on a Target or View. In your case that's recycling the GifDrawable before you start loading the next image, which breaks things.

The error messages could be improved, but if you want to keep the cross fade, see: http://bumptech.github.io/glide/doc/transitions.html#cross-fading-across-requests

**Comment by realdeadfish** [comment]
Hrm... I'd like to avoid the usage of things like `ViewSwitcher`, because I'm in a `RecyclerView` context here and the view hierarchy is already complex enough. But thanks for clearing that up. Maybe I find another solution, e.g. by accessing Glide's image cache somehow and grab the previously loaded bitmap from there.

**Comment by sjudd** [comment]
You could also probably write your own ItemAnimator to deal with the updates and cross fade between items that have changed. 

Obtaining the item from Glide's cache will result in the same problem. You'd need to make a new request to load the image into a new Target to be safe. Since you're in a scrolling list, you'd also want to make sure you cleared that Target at some point so the Bitmap can be re-used and you avoid jank. You cant however clear the Target until your animation is done and you've removed the Bitmap from wherever it's being displayed. A ViewSwitcher makes this fairly easy to think about.

**Comment by realdeadfish** [comment]
The issue is really that I want to replace an already loaded image (thumbnail) with a more resourceful version. I saw several threads about this, providing all hacky solutions that I guess introduce more problems than they actually solve.
Maybe there is a way to tackle such an issue on the library level with some special option, like "crossFadeFromPrevious()"? From what I saw while debugging through the code Glide clears up a previous image quite early in the process when a new image is loaded - maybe it's an option to postpone this up until the point where the target is actually ready to be rendered?
